lib/trivial.nix

```diff
@@ -205,6 +205,13 @@ rec {
   */
   inNixShell = builtins.getEnv "IN_NIX_SHELL" != "";
 
+  /* Determine whether the function is being called from inside pure-eval mode
+     by seeing whether `builtins` contains `currentSystem`. If not, we must be in
+     pure-eval mode. This also implies we are in a Flake.
+
+     Type: inPureEvalMode :: bool
+  */
+  inPureEvalMode = ! builtins ? currentSystem;
```

Comment:
Is it guaranteed that this check will be correct from now on? For context of what I'm talking about, please read the discussion in https://github.com/NixOS/nix/pull/5301 (tl;dr one could assume previously that `builtins ? getFlake` is sufficient to check if the experimental feature is enabled, now it throws an eval-error, but *exists*. So I'd be careful with these kinds of assumptions in general).

also cc @edolstra whom I discussed this with back then.

Reply:
It's my view that `builtins.inFlake` or `builtins.inPureEvalMode` should exist and be a primop, having looked at some of the Nix code. I would not know how to implement it, but it's the only way to guarantee that these things will continue working.

Reply:
Perhaps I'm not realizing that we're thinking the same, but let me put it in a slightly different way to be sure: `inFlake`/`inPureEvalMode` should be a primop *if* it exists.

I mean, I'm not sure if that's desirable (@edolstra raised concerns about these hacks in the PR I linked above already), so this is something to be discussed in `NixOS/nix` IMHO.

IIRC you have a global variable called `evalSettings` in `libexpr` and can access it via `evalSettings.pureEval`.

Regarding this PR: I'm :-1: on adding `inPureEvalMode`, instead we should just add the flake-specific thing to the error-message in any case for two reasons:
* We're doing it already for several other cases without checking which actually applies
* People may learn about `--impure` "on the fly" even if they don't use flakes yet. What I mean is, I think a list of all possible fixes is actually a good thing.

Reply:
I'm happy to do that too, I'll wait for everyone else's review, since this could go either way.

Reply:
> instead we should just add the flake-specific thing to the error-message in any case

I kinda prefer this option too, it is more consistent.

However the only issue I can see is that maybe this will promote Flakes and AFAIK our approach is to not divulge it yet since it is experimental. But maybe the ship sailed already (I think many `nixos-*` already has a `--flake` option that show regardless?).

Reply:
> However the only issue I can see is that maybe this will promote Flakes and AFAIK our approach is to not divulge it yet since it is experimental. But maybe the ship sailed already (I think many nixos-* already has a --flake option that show regardless Flakes is enabled or not?).

Hydra has flakes support, nixos-rebuild & other stuff as well, so this shouldn't be a big deal.

Reply:
Yeah, like I said, the ship probably shipped already.

For me, if @MatthewCroughan is fine in removing the if and just always showing the message, I would approve.